src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+
+      // 7, 9.1:
+      var attributeChangedCallback = getCallback('attributeChangedCallback');
+
+      // 8, 9.2, 9.3:
+      var observedAttributes = constructor['observedAttributes'] || [];
+
+      // 15:
+      // @type {CustomElementDefinition}
+      var definition = {
+        name: name,
+        localName: localName,
+        constructor: constructor,
+        connectedCallback: connectedCallback,
+        disconnectedCallback: disconnectedCallback,
+        attributeChangedCallback: attributeChangedCallback,
+        observedAttributes: observedAttributes,
+      };
+
+      // 16:
+      this._definitions.set(localName, definition);
+      this._constructors.set(constructor, localName);
+
+      // 17, 18, 19:
+      this._addNodes(doc.childNodes);
+
+      // 20:
+      var deferred = this._whenDefinedMap.get(localName);
+      if (deferred) {
+        deferred.resolve(undefined);
+        this._whenDefinedMap.delete(localName);
+      }
+    },
+
+    /**
+     * Returns the constructor defined for `name`, or `null`.
+     *
+     * @param {string} name
+     * @return {Function|undefined}
+     */
+    get: function(name) {
+      // https://html.spec.whatwg.org/multipage/scripting.html#custom-elements-api
+      var def = this._definitions.get(name);
+      return def ? def.constructor : undefined;
+    },
+
+    /**
+     * Returns a `Promise` that resolves when a custom element for `name` has
+     * been defined.
+     *
+     * @param {string} name
+     * @return {Promise}
+     */
+    whenDefined: function(name) {
+      // https://html.spec.whatwg.org/multipage/scripting.html#dom-customelementsregistry-whendefined
+      if (!customNameValidation.test(name)) {
+        return Promise.reject(
+          new SyntaxError(`The element name '${name}' is not valid.`));
+      }
+      if (this._definitions.has(name)) {
+        return Promise.resolve();
+      }
+      var deferred = {
+        promise: null,
+      };
+      deferred.promise = new Promise(function(resolve, _) {
+       deferred.resolve = resolve;
+      });
+      this._whenDefinedMap.set(name, deferred);
+      return deferred.promise;
+    },
+
+    /**
+     * Causes all pending mutation records to be processed, and thus all
+     * customization, upgrades and custom element reactions to be called.
+     * `enableFlush` must be true for this to work. Only use during tests!
+     */
+    flush: function() {
+      if (this.enableFlush) {
+        console.warn("flush!!!");
+        this._observers.forEach(function(observer) {
+          this._handleMutations(observer.takeRecords());
+        }, this);
+      }
+    },
+
+    _setNewInstance: function(instance) {
+      this._newInstance = instance;
+    },
+
+    /**
+     * Observes a DOM root for mutations that trigger upgrades and reactions.
+     * @private
+     */
+    _observeRoot: function(root) {
+      root.__observer = new MutationObserver(this._handleMutations.bind(this));
```

Comment:
Added an assert, since this should only be called if the root isn't already observed.